pyspark/bigdl/contrib/onnx2bigdl/load_graph.py

```diff
@@ -0,0 +1,131 @@
+#
+
+import numpy as np
+from bigdl.nn.onnx.layer import *
+from bigdl.nn.layer import Identity, Model
+from .ops_mapping import _convert_map as convert_map
+
+
+class OnnxGraph(object):
+	def __init__(self, graph_proto=None):
+		self._inputs = list()
+		self._outputs = list()
+		self._tensors = dict()
+		self._modules = dict()
+		self._root = list()
+		self._dummy_root = Identity()()
+		self._graph = self.load_graph(graph_proto)
+		return
+
+	def load_graph(self, graph_proto):
+		if not graph_proto:
+			return None
+		tensor_set = set()
+
+		for tensor in graph_proto.initializer:
+			if not tensor.name.strip():
+				raise ValueError("Tensor's name is required")
+			tensor_set.add(tensor.name)
+			tensor_data = self._parse_tensor_data(tensor)
+			self._tensors[tensor.name] = (tensor_data, tensor_data.shape)
+			
+		for gin in graph_proto.input:
+			if gin.name not in tensor_set:
+				self._inputs.append(gin.name)
+				shape = tuple([dim.dim_value for dim in gin.type.tensor_type.shape.dim])
+				self._modules[gin.name] = Identity()(self._dummy_root)
+				self._tensors[gin.name] = (None, shape)
+
+		for gout in graph_proto.output:
+			if gout.name not in tensor_set:
+				self._outputs.append(gout.name)
+
+		for node in graph_proto.node:
+			name = node.name.strip()
+			op_type = node.op_type
+			inputs = [self._tensors[n] for n in node.input]
+			outputs = node.output
+			prev_modules = [self._modules[n] for n in node.input if n not in tensor_set]
+			attrs = self._parse_node_attr(node)
+
+			if len(prev_modules) == 0:
+				self._root.append((name, op_type))
+				prev_modules = [self._dummy_root]
+
+			bigdl_module, outputs_shape = self._make_module_from_onnx_node(op_type, inputs, prev_modules, attrs, outputs)
+
+			assert len(outputs) == len(outputs_shape)
+
+			for out, out_shape in zip(outputs, outputs_shape):
+				self._modules[out] = bigdl_module
+				self._tensors[out] = (None, out_shape)
+
+		in_modules = [self._modules[m] for m in self._inputs]
+		out_modules = [self._modules[m] for m in self._outputs]
+		model = Model([self._dummy_root], out_modules)
+
+		return model
+
+	def _make_module_from_onnx_node(self, op_type, inputs, prev_modules, attrs, outputs):
+		module = None
+		out_shapes = []
+		if op_type in convert_map:
+			module, out_shapes = convert_map[op_type](inputs, prev_modules, attrs, outputs)
+		else:
+			raise NotImplemented(op_type)
+		return module, out_shapes
+
+	def _parse_tensor_data(self, tensor_proto):
+		try:
+			from onnx.numpy_helper import to_array
+		except ImportError:
+			raise ImportError("Onnx and protobuf need to be installed.")
+		if len(tuple(tensor_proto.dims)) > 0:
+			np_array = to_array(tensor_proto).reshape(tuple(tensor_proto.dims))
+		else:
+			# If it is a scalar tensor
+			np_array = np.array([to_array(tensor_proto)])
+		return np_array
+
+	def _parse_node_attr(self, node_proto):
+		attrs = {}
+		attr_proto = node_proto.attribute
+
+		for attr in attr_proto:
+			for field in ['f', 'i', 's']:
```

Comment:
what does f i and s mean respectively ?

Reply:
seems for floats, ints and strings ?

Reply:
They are float, int, string.
And there are separate logic dealing with list of them.